core/src/credential.rs

```diff
@@ -0,0 +1,147 @@
+
+#![allow(unused_variables)]
+#![allow(dead_code)]
+use std::{error::Error, time::{Instant, SystemTime}};
+
+use serde::{Deserialize, Serialize};
+use serde_json::Value;
+
+// cred_subject is a generic that implements trait X
+// trait X allows us to encode that object into JSON-LD
+// We provide types that implement trait X for the cred types that we support
+// Users can also user their own types that implement trait X if they need a different structure
+// --- 
+// Default context and Cred types are defaulted but can be redefined 
+
+type VCContext = [&'static str];
+
+pub(crate) const JSON_LD_CONTEXT:  VCContext = ["https://www.w3.org/2018/credentials/v1", "https://www.w3.org/2018/credentials/examples/v1"];
+
+pub const CRED_TYPE_PERMANENT_RESIDENT_CARD: str = "PermanentResidentCard";
+pub const CRED_TYPE_BANK_CARD: str = "BankCard";
+
+type CredentialSubject = serde_json::Value;
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct VerifiableCredential <'a> {
+    #[serde(rename = "@context")]
+	context:  VCContext,
+    #[serde(rename = "@id")]
+	id: String,
+    cred_type:  String,
+	issuer: String,
+	issuance_date: SystemTime,
+	subject: CredentialSubject,
+}
+
+pub struct IssuedCredential <'a> {
+    verifiable_credential:VerifiableCredential,
+    proof: *mut CredentialProof,
+}
+
+pub struct CredentialManager {
+    credential_subject: HashMap<&str, &str>,
+    context: VCContext,
+}
+
+trait VC {
+    fn getSubject(&self, key: &str) -> Result<String, Error> {
+        match self.credential_subject.get(key) {
+            Some(&value) => Ok(value.to_string()),
+            None => Err(Error::new(ErrorKind::Other, "Key not found")),
+        } 
+    }
+    fn getContext(&self) -> VCContext {
```

Comment:
I think what we should expect here is just a basic string array. I think requiring them to use our types isn't necessary here

Reply:
Not sure, because so far it array of contexts. But what if we change this array and decide to bind it to a certain credential type? I remember it was discussed, so far I've implemented the simplest option, but it can be a subject for a change, that's why it is a custom type. We can change the underlying type but VCContext in all functions signatures will remain the same. Just for flexibility

Reply:
Alright that's fair I can see the value of that